File: stacks/rust/base/code/database.tpl.rs
// File: database.tpl.rs
// Purpose: SQLx async database patterns with compile-time checked queries
// Generated for: {{PROJECT_NAME}}

use sqlx::{postgres::PgPoolOptions, PgPool, FromRow};
use anyhow::Result;

pub async fn create_pool(database_url: &str) -> Result<PgPool> {
    let pool = PgPoolOptions::new()
        .max_connections(5)
        .min_connections(1)
        .acquire_timeout(std::time::Duration::from_secs(30))
        .connect(database_url)
        .await?;

    Ok(pool)
}

// Example model
#[derive(Debug, FromRow)]
pub struct User {
    pub id: i32,
    pub email: String,
    pub name: Option<String>,
}

// Repository trait
#[async_trait::async_trait]
pub trait Repository<T> {
    async fn find_by_id(&self, id: i32) -> Result<Option<T>>;
    async fn find_all(&self, limit: i64, offset: i64) -> Result<Vec<T>>;
    async fn create(&self, entity: &T) -> Result<T>;
    async fn delete(&self, id: i32) -> Result<()>;
}

// User repository implementation
pub struct UserRepository {
    pool: PgPool,
}

impl UserRepository {
    pub fn new(pool: PgPool) -> Self {
        Self { pool }
    }

    pub async fn find_by_id(&self, id: i32) -> Result<Option<User>> {
        let user = sqlx::query_as!(
            User,
            "SELECT id, email, name FROM users WHERE id = $1",
            id
        )
        .fetch_optional(&self.pool)
        .await?;

        Ok(user)
    }

    pub async fn find_by_email(&self, email: &str) -> Result<Option<User>> {
        let user = sqlx::query_as!(
            User,
            "SELECT id, email, name FROM users WHERE email = $1",
            email
        )
        .fetch_optional(&self.pool)
        .await?;

        Ok(user)
    }

    pub async fn create(&self, email: &str, name: Option<&str>) -> Result<User> {
        let user = sqlx::query_as!(
            User,
            "INSERT INTO users (email, name) VALUES ($1, $2) RETURNING id, email, name",
            email,
            name
        )
        .fetch_one(&self.pool)
        .await?;

        Ok(user)
    }
}

File: stacks/python/base/code/database.tpl.py
"""
File: database.tpl.py
Purpose: SQLAlchemy 2.0 async database setup with repository pattern
Generated for: {{PROJECT_NAME}}
"""

from contextlib import asynccontextmanager
from typing import AsyncGenerator, TypeVar, Generic, Type, Optional, List

from sqlalchemy.ext.asyncio import AsyncSession, create_async_engine, async_sessionmaker
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy import select

DATABASE_URL = "postgresql+asyncpg://user:password@localhost/dbname"

engine = create_async_engine(DATABASE_URL, echo=False, pool_size=5, max_overflow=10)
async_session_factory = async_sessionmaker(engine, expire_on_commit=False)

class Base(DeclarativeBase):
    pass

@asynccontextmanager
async def get_session() -> AsyncGenerator[AsyncSession, None]:
    async with async_session_factory() as session:
        try:
            yield session
            await session.commit()
        except Exception:
            await session.rollback()
            raise

# Generic Repository Pattern
T = TypeVar("T", bound=Base)

class BaseRepository(Generic[T]):
    def __init__(self, session: AsyncSession, model: Type[T]):
        self.session = session
        self.model = model

    async def get_by_id(self, id: int) -> Optional[T]:
        result = await self.session.execute(
            select(self.model).where(self.model.id == id)
        )
        return result.scalar_one_or_none()

    async def get_all(self, limit: int = 100, offset: int = 0) -> List[T]:
        result = await self.session.execute(
            select(self.model).limit(limit).offset(offset)
        )
        return list(result.scalars().all())

    async def create(self, entity: T) -> T:
        self.session.add(entity)
        await self.session.flush()
        await self.session.refresh(entity)
        return entity

    async def delete(self, entity: T) -> None:
        await self.session.delete(entity)

# Example Model
# class User(Base):
#     __tablename__ = "users"
#     id: Mapped[int] = mapped_column(primary_key=True)
#     email: Mapped[str] = mapped_column(unique=True)

# Example Usage
# async with get_session() as session:
#     repo = BaseRepository(session, User)
#     user = await repo.get_by_id(1)
